### PR TITLE
1370 tag page wp

### PR DESCRIPTION
**Pull request by flooie**
PR for basic React and Webpack configurations.  

Part I 
Includes the compiled Javascript for tagging without the react code used to build it.
Updates the HTML to support our new React world.

### CODE REVIEW

File: cl/assets/react/@types/alltypes.d.ts
Comment (mlissner) [inline cl/assets/react/@types/alltypes.d.ts:1]: Is this still used?

**Review by mlissner — CHANGES_REQUESTED** [review]
OK, here's the first round of comments. I don't know if Github will let me do a second review, but hopefully it will.

**Comment by mlissner** [comment]
I pulled this locally and:

1. We need additions to the readme to talk about npm and whatnot. 
2. When I tried to run `npm run build`, I got some funky stuff:

```
(cl-Ib5QbsQA-py3.8) 05:16:52::mlissner@gabbro::~/Programming/courtlistener/cl 
↪ npm run build

> courtlistener-webpack-loader@1.0.0 build /home/mlissner/Programming/courtlistener/cl
> NODE_ENV=production ./node_modules/.bin/webpack --config webpack.config.js

Browserslist: caniuse-lite is outdated. Please run:
npx browserslist@latest --update-db
Hash: e1225ab5afaa3ec67826
Version: webpack 4.44.1
Time: 4894ms
Built at: 05/05/2021 5:17:00 PM
                                Asset      Size  Chunks                         Chunk Names
                              main.js  74.5 KiB    main  [emitted]              main
                           main.js.gz  12.3 KiB          [emitted]              
                          main.js.map  41.8 KiB    main  [emitted] [dev]        main
                       main.js.map.gz  10.5 KiB          [emitted]              
       vendor-a52896ea296e4577125a.js     3 MiB  vendor  [emitted] [immutable]  vendor
    vendor-a52896ea296e4577125a.js.gz   491 KiB          [emitted]              
   vendor-a52896ea296e4577125a.js.map  2.68 MiB  vendor  [emitted] [dev]        vendor
vendor-a52896ea296e4577125a.js.map.gz   461 KiB          [emitted]              
Entrypoint main = vendor-a52896ea296e4577125a.js vendor-a52896ea296e4577125a.js.map main.js main.js.map
[0] multi ./assets/react/index 28 bytes {main} [built]
[./assets/react/ListItem.tsx] 1.37 KiB {main} [built]
[./assets/react/TagList.tsx] 4.64 KiB {main} [built]
[./assets/react/TagListInner.tsx] 3.64 KiB {main} [built]
[./assets/react/TagPage.tsx] 6.66 KiB {main} [built]
[./assets/react/TagSelect.tsx] 9.56 KiB {main} [built]
[./assets/react/_fetch.ts] 2.7 KiB {main} [built]
[./assets/react/_useTags.ts] 13 KiB {main} [built]
[./assets/react/index.tsx] 1.12 KiB {main} [built]
[./node_modules/webpack/buildin/global.js] (webpack)/buildin/global.js 472 bytes {vendor} [built]
    + 548 hidden modules

ERROR in ./assets/react/TagPage.tsx
Module not found: Error: Can't resolve 'easymde/dist/easymde.min.css' in '/home/mlissner/Programming/courtlistener/cl/assets/react'
 @ ./assets/react/TagPage.tsx 7:0-38
 @ ./assets/react/index.tsx
 @ multi ./assets/react/index

ERROR in ./assets/react/TagPage.tsx
Module not found: Error: Can't resolve 'react-input-switch' in '/home/mlissner/Programming/courtlistener/cl/assets/react'
 @ ./assets/react/TagPage.tsx 9:0-40 91:67-73
 @ ./assets/react/index.tsx
 @ multi ./assets/react/index

ERROR in ./assets/react/TagListInner.tsx
Module not found: Error: Can't resolve 'react-input-switch' in '/home/mlissner/Programming/courtlistener/cl/assets/react'
 @ ./assets/react/TagListInner.tsx 5:0-40 31:42-48
 @ ./assets/react/TagList.tsx
 @ ./assets/react/index.tsx
 @ multi ./assets/react/index

ERROR in ./assets/react/TagPage.tsx
Module not found: Error: Can't resolve 'react-showdown' in '/home/mlissner/Programming/courtlistener/cl/assets/react'
 @ ./assets/react/TagPage.tsx 3:0-42 168:40-52 192:38-50
 @ ./assets/react/index.tsx
 @ multi ./assets/react/index

ERROR in ./assets/react/TagPage.tsx
Module not found: Error: Can't resolve 'react-simplemde-editor' in '/home/mlissner/Programming/courtlistener/cl/assets/react'
 @ ./assets/react/TagPage.tsx 4:0-47 203:38-47
 @ ./assets/react/index.tsx
 @ multi ./assets/react/index

ERROR in ./assets/react/TagPage.tsx
Module not found: Error: Can't resolve 'style-loader' in '/home/mlissner/Programming/courtlistener/cl'
 @ ./assets/react/TagPage.tsx 8:0-24
 @ ./assets/react/index.tsx
 @ multi ./assets/react/index
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! courtlistener-webpack-loader@1.0.0 build: `NODE_ENV=production ./node_modules/.bin/webpack --config webpack.config.js`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the courtlistener-webpack-loader@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
```

There's going to be a lot to go through. We should probably plan on cranking through a lot of this tomorrow (Thursday).

**Comment by mlissner** [comment]
Quitting for now, but my next step will be to figure out what's up with those missing dependencies.

File: cl/assets/react/@types/alltypes.d.ts
Comment (flooie) [reply cl/assets/react/@types/alltypes.d.ts:1]: yes.

**Comment by flooie** [comment]
@mlissner did you install the new dependencies in the package.json - it looks like you need to just update the dependencies on your end before you try to build with them.

File: cl/favorites/templates/tag.html
Comment (flooie) [reply cl/favorites/templates/tag.html:18]: thanks.  I don't think I want to use static_prefix either.  I think I meant to load static actually.

File: cl/favorites/views.py
Comment (flooie) [reply cl/favorites/views.py:137]: indeed

File: cl/settings/10-public.py
Comment (flooie) [reply cl/settings/10-public.py]: Correct.  Let's dump it.

File: cl/assets/react/tag-page.css
Comment (flooie) [reply cl/assets/react/tag-page.css:3]: I can do another review.  It's definitely possible that things slipped through.

**Comment by flooie** [comment]
This is ready for another review @mlissner

File: cl/donate/templates/donate.html
Comment (mlissner) [inline cl/donate/templates/donate.html:24]: These changes bring back bitcoin. They shouldn't be here.

File: cl/package.json
Comment (mlissner) [inline cl/package.json:55]: Can you go through and make sure that all of these dependencies are in the right place? For example, do we need types in our dependencies or in our dev deps? My assumption is we can lighten the JS we ship by checking this a little more?

File: cl/package.json
Comment (mlissner) [inline cl/package.json:20]: Can you just skim this list for anything that you don't recognize and purge it? I know I installed a bunch of things along the way that I thought might be useful, but I don't know if I was good about actually removing them again.

File: cl/opinion_page/templates/docket_tabs.html
Comment (mlissner) [inline cl/opinion_page/templates/docket_tabs.html:93]: Sorry, if I missed it, but I can't find where this is being used anywhere. Is it possible we don't use it? If we don't, let's remove it from here and elsewhere. Are all the others in use wherever we have react-root?

File: cl/opinion_page/templates/docket_tabs.html
Comment (mlissner) [inline cl/opinion_page/templates/docket_tabs.html:95]: It's a bit weird we have this here, right?

File: cl/settings/10-public.py
Comment (mlissner) [inline cl/settings/10-public.py:286]: This looks like it was added by Nick. It's a bit weird because we already have `BASE_DIR` configured in settings.py. Can you try removing it and make sure everything still works? I'm not sure what this variable is supposed to do, honestly, but it's not one that we usually would need to tinker with.

File: cl/assets/react/TagList.tsx
Comment (mlissner) [inline cl/assets/react/TagList.tsx:17]: That's too bad. Best time to update it is now though. Have you looked at how hard it is?

**Review by mlissner — COMMENTED** [review]
Well, I wasn't able to finish all the ts and react stuff, but here are some initial comments to chew on. Small stuff so far. I'm not finding much in react or TS since I'm not super good at it, but perhaps I'll find something of use to us. More on Monday!

File: cl/assets/react/TagList.tsx
Comment (litewarp) [reply cl/assets/react/TagList.tsx:17]: Shouldn't be too hard to adopt the new api. See, https://react-query.tanstack.com/guides/paginated-queries. It looks like the only thing extra that's needed is to pass the "keepPreviousData" option to true.

File: cl/settings/10-public.py
Comment (litewarp) [reply cl/settings/10-public.py:286]: The only use of this would have been to point the webpack loader to the correct directory. So if that is sorted out, you could probably delete with no effect on anything else.

File: cl/package.json
Comment (litewarp) [reply cl/package.json:55]: So it shouldn't matter where the packages are placed. The way webpack works is that it will (or should) only bundle imported modules (often referred to as "tree-shaking"). Meaning that even if a package is in "Dependencies" and not dev dependencies, the only affect it will have will be on the server at build time (i.e., deployment). Once the JS is compiled and minified, nothing else is added into it.

You can see what the bundle sizes are by running the webpack-bundle-analyzer.

File: cl/assets/react/TagSelect.tsx
Comment (mlissner) [inline cl/assets/react/TagSelect.tsx:81]: Can you please do something better here? Using this selector is fragile, and the variable name is unclear.

File: cl/assets/react/tag-page.css
Comment (mlissner) [inline cl/assets/react/tag-page.css:2]: Can you skim this, Bill, and see if anything looks obsolete to you?

**Comment by mlissner** [comment]
Well, I took a look through the typescript and didn't see much I could improve. I wish I had more to give you here, but I'm not that good at this kind of code. @litewarp says he'll take a look too. We should wait for that, fix the few things I've found, and then land this, I guess.

**Comment by flooie** [comment]
I’m gonna take that as a compliment.

**Comment by mlissner** [comment]
One other to do, @flooie:

 - [ ] Add documentation for NPM etc. to DEVELOPING.md

File: cl/assets/react/TagSelect.tsx
Comment (flooie) [reply cl/assets/react/TagSelect.tsx]: yes.  yes it is ugly

File: cl/opinion_page/templates/docket_tabs.html
Comment (flooie) [reply cl/opinion_page/templates/docket_tabs.html:95]: which part do you think is weird?

Comment (flooie) [reply cl/opinion_page/templates/docket_tabs.html:95]: this is how I elegantly pass the data to the React side

File: cl/package.json
Comment (flooie) [reply cl/package.json:55]: @mlissner what @litewarp said is what I also see.  Although I did try to purge anything that we weren't using, I did not touch anything like Webpack-cli

Comment (mlissner) [reply cl/package.json:55]: Makes sense regarding bundling. I didn't mean to just highlight the webpack stuff though, Bill. Could you just eye the list of all packages and see if there are ones to remove?

File: cl/assets/react/TagList.tsx
Comment (flooie) [reply cl/assets/react/TagList.tsx]: This should be resolved.  thanks for the tip @litewarp

File: cl/assets/react/TagSelect.tsx
Comment (flooie) [reply cl/assets/react/TagSelect.tsx]: √

File: cl/assets/react/tag-page.css
Comment (flooie) [reply cl/assets/react/tag-page.css]: yes. good eye

File: cl/settings/10-public.py
Comment (flooie) [reply cl/settings/10-public.py]: I took it out and dont see any issues building from Webpack.

File: cl/package.json
Comment (flooie) [reply cl/package.json:55]: absolutely

Comment (flooie) [reply cl/package.json:55]: two dependencies look superfluous to me - and to a dep check `i` and `caniuse`

File: cl/assets/react/tag-page.css
Comment (flooie) [reply cl/assets/react/tag-page.css]: √

File: cl/opinion_page/templates/docket_tabs.html
Comment (flooie) [reply cl/opinion_page/templates/docket_tabs.html:93]: In this instance -- 

The data attributes get automatically transferred and converted to camel case format.

`const TagSelect: React.FC<UserState> = ({ userId, userName, editUrl, docket }) => {
`

File: cl/donate/templates/donate.html
Comment (flooie) [reply cl/donate/templates/donate.html:24]: this is a bad rebase I think.  I got mixed up about removing bitcoin no taking it away.

File: cl/favorites/urls.py
Comment (flooie) [reply cl/favorites/urls.py:1]: yes

**Comment by mlissner** [comment]
It's time to merge this danged thing. We can keep polishing it endlessly, but it's damned good and I'm proud of where it is. I'm going to do a quick help page, I think, and then deploy this monumental lift. Thank you again, @flooie for scraping this off the floor where I left it and making it into a complete and ready thing. Thank you again, @litewarp for all your help getting this going and writing so much code for it. I don't think we'd have gotten it done without you. This is hella cool.

**Pull request merged**
